src/nvim/api/buffer.c

```diff
@@ -453,6 +453,11 @@ void nvim_buf_set_lines(uint64_t channel_id,
   aco_save_T aco;
   aucmd_prepbuf(&aco, (buf_T *)buf);
 
+  if (!MODIFIABLE(buf)) {
+    api_set_error(err, kErrorTypeException, "Buffer is not 'modifiable'");
+    goto end;
+  }
+
   if (u_save((linenr_T)(start - 1), (linenr_T)end) == FAIL) {
     api_set_error(err, kErrorTypeException, "Failed to save undo information");
```

Comment:
Before this change, the error message would be `Failed to save undo information`.

Reply:
With `setbufline()` the error is:

> E21: Cannot make changes, 'modifiable' is off

Reply:
Yes, I intentionally chose something that mentions "buffer". 

Anyways, API messages in general need to be revisited. 

1. it would be nice to avoid duplicate validation when the editor already handles it (EXX messages)
2. there should be more consistency/uniformity/standardization in the API messages 
3. a "validation module" could avoid lots of redundant code in eval.c and the API